xds/src/main/java/io/grpc/xds/XdsServerWrapper.java

```diff
@@ -0,0 +1,439 @@
+import java.util.EnumSet;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+final class XdsServerWrapper extends Server {
+  private static final Logger logger = Logger.getLogger(XdsServerWrapper.class.getName());
+
+  private final SynchronizationContext syncContext = new SynchronizationContext(
+      new Thread.UncaughtExceptionHandler() {
+        @Override
+        public void uncaughtException(Thread t, Throwable e) {
+          logger.log(Level.SEVERE, "Exception!" + e);
+          // TODO(chengyuanzhang): implement cleanup.
+        }
+      });
+
+  private final String listenerAddress;
+  private final ServerBuilder<?> delegateBuilder;
+  private ScheduledExecutorService timeService = SharedResourceHolder.get(GrpcUtil.TIMER_SERVICE);
+  private final long retryDelayNano;
+  private final XdsClientPoolFactory xdsClientPoolFactory;
+  private final XdsServingStatusListener listener;
+  private final AtomicReference<FilterChainSelector> filterChainSelectorRef;
+  private final AtomicBoolean started = new AtomicBoolean(false);
+  private final AtomicBoolean shutdown = new AtomicBoolean(false);
+  private boolean isServing;
+  private final CountDownLatch internalTerminationLatch = new CountDownLatch(1);
+  private final SettableFuture<IOException> initialStartFuture = SettableFuture.create();
+  private boolean initialStarted;
+  private ScheduledHandle restartTimer;
+  private ObjectPool<XdsClient> xdsClientPool;
+  private XdsClient xdsClient;
+  private DiscoveryState discoveryState;
+  private volatile Server delegate;
+
+  XdsServerWrapper(
+      String listenerAddress,
+      ServerBuilder<?> delegateBuilder,
+      long retryDelayNano,
+      XdsServingStatusListener listener,
+      AtomicReference<FilterChainSelector> filterChainSelectorRef,
+      XdsClientPoolFactory xdsClientPoolFactory) {
+    this.listenerAddress = checkNotNull(listenerAddress, "listenerAddress");
+    this.delegateBuilder = checkNotNull(delegateBuilder, "delegateBuilder");
+    this.retryDelayNano = retryDelayNano;
+    this.listener = checkNotNull(listener, "listener");
+    this.filterChainSelectorRef = checkNotNull(filterChainSelectorRef, "filterChainSelectorRef");
+    this.xdsClientPoolFactory = checkNotNull(xdsClientPoolFactory, "xdsClientPoolFactory");
+    this.delegate = delegateBuilder.build();
+  }
+
+  @Override
+  public Server start() throws IOException {
+    checkState(started.compareAndSet(false, true), "Already started");
+    syncContext.execute(new Runnable() {
+      @Override
+      public void run() {
+        internalStart();
+      }
+    });
+    syncContext.drain();
```

Comment:
`syncContext.execute()` includes a call to `drain()` . So I am trying to understand the purpose of this extra call...